src/core/shutdown.c

```diff
@@ -146,6 +148,23 @@ static int parse_argv(int argc, char *argv[]) {
 }
 
 static int switch_root_initramfs(void) {
+        struct statfs st;
+        int r;
+
+        r = statfs("/run", &st);
+        if (r < 0)
+                log_error_errno(errno, "Failed to determine mount flags for /run: %m");
+
+        /* If /run is mounted with NOEXEC, remount without it, as otherwise exec /shutdown post switch_root will fail */
+        if (r == 0 && (st.f_flags & ST_NOEXEC)) {
+                if (mac_smack_use())
+                        r = mount(NULL, "/run", NULL, MS_REMOUNT|MS_NOSUID|MS_NODEV|MS_STRICTATIME, "mode=755,smackfsroot=*");
+                else
+                        r = mount(NULL, "/run", NULL, MS_REMOUNT|MS_NOSUID|MS_NODEV|MS_STRICTATIME, "mode=755");
```

Comment:
Shouldn't this be `st.f_flags & ~ST_NOEXEC | MS_REMOUNT`? Also, I'd prefer to use tertiary operator just for the third argument: `mac_smack_use() ? "mode=755,smackfsroot=*" : "mode=755".

Reply:
Ooooh, I like both of your suggestions!